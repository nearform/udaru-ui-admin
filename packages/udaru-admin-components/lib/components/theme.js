// Adapted from default theme on https://github.com/jxnblk/styled-system

// breakpoint values
// any array length works with styled-system
const breakpoints = ['500px', '800px', '1200px']

const colors = {
  primary: '#337ab7',
  primaryActive: '#02A171',
  danger: '#d9534f',
  lightGrey: '#D8D8D8',
  lightGrey2: '#888',
  lightGrey3: '#eee',
  darkGrey: '#555',
  black: '#333',
  green: '#5cb85c',
  lightBlue: '#5bc0de',
  orange: '#f0ad4e'
}

export { colors }

export const radius = '4px'

export const maxContainerWidth = '1280px'

// boxShadows
export const boxShadows = [
  `0 0 2px 0 rgba(0,0,0,.08),0 1px 4px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 2px 8px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 4px 16px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 8px 32px 0 rgba(0,0,0,.16)`
]

export const speeds = {
  slow: '1s',
  medium: '.4s',
  fast: '.2s'
}

export const easings = {
  easeOut: 'cubic-bezier(0,.6,.3,1)',
  expoEaseOut: 'cubic-bezier(0,1,.3,1)',
  bounce: 'cubic-bezier(0,1.2,.5,1.4)'
}

// space is used for margin and padding scales
// it's recommended to use powers of two to ensure alignment
// when used in nested elements
// numbers are converted to px
const space = [0, 4, 8, 16, 32, 64, 128, 256, 512]

// typographic scale
const fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 96, 128]

// for any scale, either array or objects will work
const lineHeights = [1, 1.125, 1.25, 1.5]

const fontWeights = {
  normal: 500,
  bold: 700
}

const letterSpacings = {
  normal: 'normal',
  caps: '0.25em'
}

// border-radius
const radii = [0, 2, 4, 8]

const borders = [0, '1px solid', '2px solid']

const shadows = [`0 1px 2px 0 ${colors.text}`, `0 1px 4px 0 ${colors.text}`]

const theme = {
  breakpoints,
  colors,
  space,
  fontSizes,
  lineHeights,
  fontWeights,
  letterSpacings,
  radii,
  radius,
  borders,
  shadows,
  boxShadows,
  speeds,
  easings
}

export default theme
