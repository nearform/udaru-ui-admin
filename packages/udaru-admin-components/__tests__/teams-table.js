import React from 'react'
import renderer from 'react-test-renderer'
import TeamsTable from 'teams-table'
import { createSerializer } from 'jest-emotion'
import * as emotion from 'emotion'
import theme from '../lib/components/theme'

expect.addSnapshotSerializer(createSerializer(emotion))

jest.mock('react-bootstrap-table', () => {
  return {
    BootstrapTable: props => {
      return (
        <table>
          {' // mocked table output'}
          {props.children}
        </table>
      )
    },
    TableHeaderColumn: props => {
      return (
        <tr>
          <th>{props.dataField + ' // mocked table header column'}</th>
        </tr>
      )
    }
  }
})

it('should render with props', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {
              data: [],
              total: 0
            }
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  await instance.componentDidMount()
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  global.fetch.mockRestore()
})

it('should render with error state', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: false,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance

  await instance.componentDidMount()

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  global.fetch.mockRestore()
})

it('should render create team', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  await instance.componentDidMount()

  expect(instance.state.view).toBe('LIST')
  instance.onCreate()
  expect(instance.state.view).toBe('CREATE')

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  global.fetch.mockRestore()
})

it('should render view team', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance

  expect(instance.state.view).toBe('LIST')
  await instance.setState({
    selectedRow: {
      id: '1'
    }
  })
  await instance.onView()
  expect(instance.state.view).toBe('READ')

  global.fetch.mockRestore()
})

it('should render view team', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance

  expect(instance.state.view).toBe('LIST')

  await instance.onNestedView('1', '2')

  expect(instance.state.view).toBe('READ')
  expect(instance.state.viewId).toBe('2')
  expect(instance.state.parentTeamId).toBe('1')

  global.fetch.mockRestore()
})

it('should render view parent team', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance

  expect(instance.state.view).toBe('LIST')

  await instance.onViewParent('1')

  expect(instance.state.view).toBe('READ')
  expect(instance.state.viewId).toBe('1')
  expect(instance.state.parentTeamId).toBeNull()

  global.fetch.mockRestore()
})

it('should render update team', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  await instance.componentDidMount()

  expect(instance.state.view).toBe('LIST')

  instance.setState({
    selectedRow: {
      id: '1',
      name: 'Team Name',
      description: 'Team Description'
    }
  })
  instance.onUpdate()

  expect(instance.state.view).toBe('UPDATE')

  const json = component.toJSON()

  expect(json).toMatchSnapshot()

  global.fetch.mockRestore()
})

it('should render delete team', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  await instance.componentDidMount()

  expect(instance.state.view).toBe('LIST')

  instance.setState({
    selectedRow: {
      id: '1',
      name: 'Team Name',
      description: 'Team Description'
    }
  })

  await instance.onDelete()
  expect(instance.state.view).toBe('DELETE')

  const json = component.toJSON()
  expect(json).toMatchSnapshot()

  global.fetch.mockRestore()
})

it('should render error in unknown state', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  await component.getInstance().componentDidMount()

  expect(instance.state.view).toBe('LIST')

  await instance.setState({
    view: 'IDONTKNOWTHISSTATE'
  })

  const json = component.toJSON()
  expect(json).toMatchSnapshot()

  global.fetch.mockRestore()
})

it('should render create and then go back to list and clear selectedRow', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  expect(instance.state.view).toBe('LIST')

  instance.setState({
    selectedRow: {
      id: '1',
      name: 'Team Name',
      description: 'Team Description'
    }
  })
  instance.onCreate()
  expect(instance.state.view).toBe('CREATE')
  await instance.onCancel()

  expect(instance.state.view).toBe('LIST')
  expect(instance.state.selectedRow).toBeNull()
  global.fetch.mockRestore()
})

it('should reset search change if text is blank', () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(<TeamsTable theme={theme} />)
  const instance = component.root.instance
  instance.setState({
    originalData: [{ id: '1' }],
    originalDataTotalSize: 1
  })

  instance.onSearchChange('')
  expect(instance.state.data).toEqual([{ id: '1' }])
  expect(instance.state.dataTotalSize).toEqual(1)

  global.fetch.mockRestore()
})

it('should search for value', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {
              data: [{ id: '2' }],
              total: 1
            }
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance

  await instance.onSearchChange('team')

  expect(instance.state.data).toEqual([{ id: '2' }])
  expect(instance.state.dataTotalSize).toEqual(1)

  global.fetch.mockRestore()
})

it('should search handle malformed response', () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance

  instance.onSearchChange('team')

  expect(instance.state.data).toEqual([])
  expect(instance.state.dataTotalSize).toEqual(0)

  global.fetch.mockRestore()
})

it('should handle error response', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: false,
          json: function() {
            return {}
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance
  const data = [{ id: 1 }]
  const totalSize = 1

  instance.setState({
    originalData: data,
    originalDataTotalSize: totalSize
  })

  await instance.onSearchChange('team')

  expect(instance.state.data).toEqual(data)
  expect(instance.state.dataTotalSize).toEqual(totalSize)

  global.fetch.mockRestore()
})

it('should handle select row', () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {
              data: [{ id: '2' }],
              total: 1
            }
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance

  instance.onSelect('row-1')
  expect(instance.state.selectedRow).toBe('row-1')

  global.fetch.mockRestore()
})

it('should handle page change', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {
              data: [{ id: '2' }],
              total: 1
            }
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance

  await instance.onPageChange(1, 1)

  expect(global.fetch.mock.calls[1][0]).toBe(
    '/authorization/teams?page=1&limit=1'
  )

  global.fetch.mockRestore()
})

it('should handle page list change', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {
              data: [{ id: '2' }],
              total: 1
            }
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance

  await instance.onSizePerPageList(22)

  expect(global.fetch.mock.calls[1][0]).toBe(
    '/authorization/teams?page=1&limit=22'
  )

  global.fetch.mockRestore()
})

it('should reject all running promises', async () => {
  global.fetch = jest.fn().mockImplementation(
    () =>
      new Promise((resolve, reject) => {
        resolve({
          ok: true,
          json: function() {
            return {
              data: [{ id: '2' }],
              total: 1
            }
          }
        })
      })
  )

  const component = renderer.create(
    <TeamsTable
      theme={theme}
      authorization={'my-authorization'}
      org={'my-org'}
    />
  )
  const instance = component.root.instance
  await component.unmount()

  const { _runningPromises } = instance

  expect(_runningPromises.every(p => p.hasCanceled())).toBeTruthy()

  global.fetch.mockRestore()
})
