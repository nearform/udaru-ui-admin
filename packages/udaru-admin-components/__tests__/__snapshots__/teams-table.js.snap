// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render create team 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <div
              className="page-header"
            >
              <h1>
                Create Team
              </h1>
            </div>
          </div>
          <div
            className="col-xs-12"
          />
          <div
            className="col-xs-12"
          >
            <form
              className=""
              onSubmit={[Function]}
            >
              <div
                className="form-group"
              >
                <h3>
                  Team Details
                </h3>
              </div>
              <div
                className="form-group has-feedback"
              >
                <label
                  className="control-label"
                  htmlFor="id"
                >
                  ID
                </label>
                <input
                  className="form-control"
                  id="id"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="ID"
                  type="text"
                  value=""
                />
              </div>
              <div
                className="form-group has-feedback"
              >
                <label
                  className="control-label"
                  htmlFor="name"
                >
                  Name
                </label>
                <input
                  className="form-control"
                  id="name"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Name"
                  type="text"
                  value=""
                />
              </div>
              <div
                className="form-group has-feedback"
              >
                <label
                  className="control-label"
                  htmlFor="description"
                >
                  Description
                </label>
                <input
                  className="form-control"
                  id="description"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Description"
                  type="text"
                  value=""
                />
              </div>
              <div
                className="form-group"
              >
                <h3>
                  Default Admin User
                </h3>
              </div>
              <div
                className="form-group has-feedback"
              >
                <label
                  className="control-label"
                  htmlFor="userId"
                >
                  User ID
                </label>
                <input
                  className="form-control"
                  id="userId"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="User ID"
                  type="text"
                  value=""
                />
              </div>
              <div
                className="form-group has-feedback"
              >
                <label
                  className="control-label"
                  htmlFor="userName"
                >
                  User Name
                </label>
                <input
                  className="form-control"
                  id="userName"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="User Name"
                  type="text"
                  value=""
                />
              </div>
              <button
                className="btn btn-primary"
                disabled={false}
                type="submit"
              >
                Submit
              </button>
              <button
                className="btn btn-link"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render delete team 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="page-header"
      >
        <h1>
          Delete Team
        </h1>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <h3>
        Are you sure you want to delete 
        Team Name
        ?
      </h3>
    </div>
    <div
      className="col-xs-12"
    >
      <p>
        This action 
        <strong>
          CANNOT
        </strong>
         be undone.
      </p>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <button
        className="btn btn-danger"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        DELETE
      </button>
      <button
        className="btn btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        CANCEL
      </button>
    </div>
  </div>
</div>
`;

exports[`should render error in unknown state 1`] = `
<h1>
  There was an error fetching teams.
</h1>
`;

exports[`should render update team 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <div
        className="page-header"
      >
        <h1>
          Update Team
        </h1>
      </div>
    </div>
    <div
      className="col-xs-12"
    />
    <div
      className="col-xs-12"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor={undefined}
          >
            ID
          </label>
          <input
            className="form-control"
            disabled={true}
            id={undefined}
            placeholder="ID"
            type="text"
            value="1"
          />
        </div>
        <div
          className="form-group has-feedback has-success"
        >
          <label
            className="control-label"
            htmlFor="name"
          >
            Name
          </label>
          <input
            className="form-control"
            id="name"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value="Team Name"
          />
          <span
            className="form-control-feedback glyphicon glyphicon-ok"
          />
        </div>
        <div
          className="form-group has-feedback has-success"
        >
          <label
            className="control-label"
            htmlFor="description"
          >
            Description
          </label>
          <input
            className="form-control"
            id="description"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Description"
            type="text"
            value="Team Description"
          />
          <span
            className="form-control-feedback glyphicon glyphicon-ok"
          />
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`should render with error state 1`] = `
<h1>
  There was an error fetching teams.
</h1>
`;

exports[`should render with props 1`] = `
Array [
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-xs-12"
      >
        <div
          className="btn-toolbar"
          role="toolbar"
        >
          <button
            className="btn btn-success"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            New Team
          </button>
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            style={
              Object {
                "display": "inline-block",
                "padding": "0 10px",
              }
            }
          >
            <button
              className="btn btn-primary"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "pointerEvents": "none",
                }
              }
              type="button"
            >
              View Team
            </button>
          </div>
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            style={
              Object {
                "display": "inline-block",
                "padding": "0 10px 0 0",
              }
            }
          >
            <button
              className="btn btn-info"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "pointerEvents": "none",
                }
              }
              type="button"
            >
              Update Team
            </button>
          </div>
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <button
              className="btn btn-danger"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "pointerEvents": "none",
                }
              }
              type="button"
            >
              Delete Team
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    style={
      Object {
        "margin": "30px 0",
      }
    }
  />,
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-xs-12"
      >
        <table>
           // mocked table output
          <tr>
            <th>
              id // mocked table header column
            </th>
          </tr>
          <tr>
            <th>
              name // mocked table header column
            </th>
          </tr>
          <tr>
            <th>
              description // mocked table header column
            </th>
          </tr>
          <tr>
            <th>
              usersCount // mocked table header column
            </th>
          </tr>
        </table>
      </div>
    </div>
  </div>,
]
`;
